src/lib.rs

```diff
@@ -1,6 +1,159 @@
+use std::io::Cursor;
+use std::io::prelude::*;
+
+const HDR_LEN: usize = 40;
+const TOC_BLOCK_LEN: usize = 16;
+
+type Endian = LittleEndian;
+
+#[cfg(unix)]
+fn get_process_id() -> i32 {
+    nix::unistd::getpid()
+}
+
+#[cfg(windows)]
+fn get_process_id() -> i32 {
+    unsafe { kernel32::GetCurrentProcessId() as i32 }
+}
+ 
+bitflags! {
+    /// Flags used to modify how a client exports metrics
+    pub struct MMVFlags: u32 {
+        /// Metric names aren't prefixed with MMV filename
+        const NOPREFIX = 1;
+        /// PID check is needed
+        const PROCESS  = 2;
+        /// Allow for "no value available" values
+        const SENTINEL = 4;
+    }
+}
+
+/// Client used to export metrics
+pub struct Client {
+    path: String,
+    flags: MMVFlags,
+    cluster_id: u32,
+}
+
+impl Client {
+    /// Creates a new client with default flags and cluster ID `0`
+    pub fn new(path: &str) -> Client {
+        Client::new_custom(path, MMVFlags::empty(), 0)
+    }
+
+    /// Creates a new client with flags and cluster ID
+    pub fn new_custom(path: &str, flags: MMVFlags, cluster_id: u32) -> Client {
+        Client {
+            path: path.to_owned(),
+            flags: flags,
+            cluster_id: cluster_id,
+        }
     }
+
+    /// Exports metrics by writing to an MMV file
+    pub fn export(&self) -> std::io::Result<()> {
+        let mut file = OpenOptions::new().read(true).write(true).create(true)
+            .open(&self.path)?;
+        let mmv_size = HDR_LEN + 2*TOC_BLOCK_LEN;
+        file.write(&vec![0; mmv_size])?;
+
+        let mut mmap = Mmap::open(&file, Protection::ReadWrite)?;
+        let mut cursor = Cursor::new(unsafe { mmap.as_mut_slice() });
+
+        // MMV\0
+        cursor.write_all(CString::new("MMV")?.to_bytes_with_nul())?;
+        // version
+        cursor.write_u32::<Endian>(1)?;
+        // generation1
+        let gen = time::now().to_timespec().sec;
+        cursor.write_i64::<Endian>(gen)?;
+        let gen2pos = cursor.position();
+        cursor.write_i64::<Endian>(0)?;
+        // no. of toc blocks
+        cursor.write_i32::<Endian>(2)?;
+        // flags
+        cursor.write_u32::<Endian>(self.flags.bits())?;
+        // pid
+        cursor.write_i32::<Endian>(get_process_id())?;
+        // cluster id
+        cursor.write_u32::<Endian>(self.cluster_id)?;
+
+        // metrics TOC block
+        // section type
+        cursor.write_u32::<Endian>(3)?;
+        // no. of entries
+        cursor.write_u32::<Endian>(0)?;
+        // section offset
+        cursor.write_u64::<Endian>(0)?;
+
+        // values TOC block
+        // section type
+        cursor.write_u32::<Endian>(4)?;
+        // no. of entries
+        cursor.write_u32::<Endian>(0)?;
+        // section offset
+        cursor.write_u64::<Endian>(0)?;
+
+        // unlock header
+        cursor.set_position(gen2pos);
+        cursor.write_i64::<Endian>(gen)
+    }
+}
+
+#[test]
+fn test_mmv_header() {
```

Comment:
Again, this feels like we are testing "how" MMV files are generated, instead of "what" is generated. But this can go, as a test for just writing the header. In the long term, to test the conformance of full MMV files, we should start looking into implementing MMVDump (my recommendation), or use the C or Go versions. Also, as far as testing MMVDump is concerned, you are right, ideally we would have a bunch of `input` and `output.golden` files, and we would generate `output` using our implementation and see if it exactly matches `output.golden`